libbeat/publisher/queue/diskqueue/core_loop.go

```diff
@@ -151,37 +152,19 @@ func (dq *diskQueue) handleReaderLoopResponse(response readerLoopResponse) {
 
 	// Advance the frame / offset based on what was just completed.
 	dq.segments.nextReadFrameID += frameID(response.frameCount)
-	dq.segments.nextReadOffset += segmentOffset(response.byteCount)
+	dq.segments.nextReadPosition += response.byteCount
 
-	var segment *queueSegment
-	if len(dq.segments.reading) > 0 {
-		// A segment is finished if we have read all the data, or
-		// the read response reports an error.
-		// Segments in the reading list have been completely written,
-		// so we can rely on their endOffset field to determine their size.
-		segment = dq.segments.reading[0]
-		if dq.segments.nextReadOffset >= segment.endOffset || response.err != nil {
-			dq.segments.reading = dq.segments.reading[1:]
-			dq.segments.acking = append(dq.segments.acking, segment)
-			dq.segments.nextReadOffset = 0
-		}
-	} else {
-		// A segment in the writing list can't be finished writing,
-		// so we don't check the endOffset.
-		segment = dq.segments.writing[0]
-		if response.err != nil {
-			// Errors reading a writing segment are awkward since we can't discard
-			// them until the writer loop is done with them. Instead we just seek
-			// to the end of the current data region. If we're lucky this lets us
-			// skip the intervening errors; if not, the segment will be cleaned up
-			// after the writer loop is done with it.
-			dq.segments.nextReadOffset = segment.endOffset
-		}
-	}
+	segment := dq.segments.readingSegment()
 	segment.framesRead += response.frameCount
-
-	// If there was an error, report it.
 	if response.err != nil {
+		// If there's an error, we advance to the end of the current segment.
+		// If the segment is in the reading list, it will be removed on the
+		// next call to maybeReadPending.
+		// If the segment is still in the writing list, we can't discard it
+		// until the writer loop is done with it, but we can hope that advancing
+		// to the current write position will get us out of our error state.
+		dq.segments.nextReadPosition = segment.byteCount
```

Comment:
Given the way readers and writers are coordinated, an error in a segment that is currently written would likely indicate a bug in the business logic, a bug in the framing done by the writer, or an unexpected race condition. The system will most likely recover once we did start a new segment file. In this case we might want to use `logger.Criticalf`, in order to encourage users to report bugs.

Errors on older, already closed segment files might indicate a broken/invalid segment file, or a bug in the framing. The former is to be expected if the system was not shutdown cleanly. Error level would be enough I think.